Show end-of-list marker for PostTypeList

jonathansadowski opened the pull request:
This PR adds end-of-list markers to PostTypeList:

![image](https://user-images.githubusercontent.com/363749/32191284-80a93cc4-bd7e-11e7-9821-8b16187fa2cc.png)

When the "All My Sites" view of the PostTypeList contains more than 400 posts, this PR also adds a message informing you how to view the remainder of your posts.

![image](https://user-images.githubusercontent.com/363749/32191338-b964df28-bd7e-11e7-9c5a-92adac8ac422.png)

To test:
* Apply PR.
* `ENABLE_FEATURES=posts/post-type-list npm start`.
* Visit your "Blog Posts" for a single site.
* Make sure at the end of the list (once all pages have loaded) you see the end of list marker.
* With a user that has fewer than 400 total blog posts, enter "All My Sites" mode. Scroll to the bottom of "Blog Posts" and verify the end of list marker is present, but the message about viewing the remainder of posts is not.
* With a user that has more than 400 total blog posts, , enter "All My Sites" mode. Scroll to the bottom of "Blog Posts" and verify both the end of list marker and the message about viewing the remainder of posts are present.
* Verify that clicking on "switch to a specific site" opens the site picker.

nylen commented on client/my-sites/post-type-list/max-pages-notice.jsx:
```diff
@@ -0,0 +1,65 @@
+			<div className="post-type-list__max-pages-notice">
+				{ translate(
+					'%(displayedPosts)d of %(totalPosts)d posts shown' +
+						'{{br/}}to view more posts, {{a}}switch to a specific site{{/a}}.',
```
This probably needs a singular and plural form, even though in English we would always use the plural form.  See https://github.com/Automattic/i18n-calypso#pluralization:

> It might seem redundant, but it is necessary for languages where a singular version may be used for counts other than 1.

Also, each line of this message should be a complete sentence, with punctuation and capital letters (where applicable).

nylen replied:
I got a suggestion from the i18n team here ( thanks @yoavf ) which I think is a good one:

> Showing %d post[s] out of %d

In code, it would look something like this:

```js
{ translate(
    'Showing %(displayedPosts)d post out of %(totalPosts)d.',
    'Showing %(displayedPosts)d posts out of %(totalPosts)d.',
    // ... arguments
) }
<br />
{ translate(
    'To view more posts, {{a}}switch to a specific site{{/a}}.',
    // ... arguments
) }
```

nylen commented on client/my-sites/post-type-list/index.jsx:
```diff
@@ -201,5 +237,6 @@ export default connect( ( state, ownProps ) => {
 		posts: getSitePostsForQueryIgnoringPage( state, siteId, ownProps.query ),
 		isRequestingPosts: isRequestingSitePostsForQueryIgnoringPage( state, siteId, ownProps.query ),
 		lastPage,
+		totalPosts: getSitePostsFoundForQuery( state, siteId, ownProps.query ),
```
While we are making changes here, how about making the form of all these props consistent:

```
lastPage: getSitePostsLastPageForQuery( state, siteId, ownProps.query ),
```

nylen replied:
As #18998 is currently written, `getSitePostsFoundForQuery` will no longer work in all-sites mode.  Let's make sure we fix that, whether here or in that PR (most likely by renaming this selector to `getPostsFoundForQuery` and making it work in all-sites mode too).

nylen commented on client/my-sites/post-type-list/index.jsx:
```diff
@@ -122,6 +132,18 @@ class PostTypeList extends Component {
+	isTruncatedResults() {
+		const { siteId, lastPage } = this.props;
+		return null === siteId && this.hasListFullyLoaded() && lastPage > MAX_ALL_SITES_PAGES;
+	}
```
Let's pull the contents of this method into `renderMaxPagesNotice` - it's only used once, and I'm not crazy about the name anyway.

nylen replied:
It will probably also be useful to pull out some logic to get the maximum page we should actually request (`siteId === null ? MAX_ALL_SITES_PAGES : lastPage`).  Then, we need to make sure we never actually request more than this number of pages, and show the max pages notice after loading the 10th page.

nylen commented:
This is mostly working as intended, and the code looks good so far.  However, we still try to request all of the pages (> 10) in all-sites mode, and the `ListEnd` and count notice only appear once all these requests are done:

![2017-10-26t17 25 21 0800](https://user-images.githubusercontent.com/227022/32045528-1c77fb7c-ba73-11e7-953c-32d2cb3bfc02.png)

This screenshot also shows the design issues (border around entire `PostTypeList`) blocking this PR from shipping.  As previously discussed, this is just a preliminary review, so I'll leave the PR status alone.

jonathansadowski commented:
Thanks for the review -- I've updated the PR such that it no longer attempts to query beyond 10 pages in all sites mode and addressed some of the other feedback.

nylen commented:
This is better, but still sub-optimal.  When we have 30 pages in all-sites mode, we will render 10 pages, re-render 20 times due to `setState( { maxRequestedPage: maxRequestedPage + 1 } )`, then finally show the notice.  We can simplify a bit further still - I'll give it a try.

nylen commented:
I think this is ready to go.  Leaving in "Needs Review" so that someone else can verify my changes.

Minor annoyance:  when you click "switch to a specific site", then click in the main area to dismiss the site picker, the list scrolls back to the top.  This is not related to the changes here and should be fixed separately.

jonathansadowski commented:
Thanks for the review!
